File: 242_valid_anagram/hash.py
class Solution:
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        """
        Complexity: O(n)
        Space: O(1) the table size stay constant no matter how large the n is
        """
        occurrence = {}
        for c in s:
            if c in occurrence:
                occurrence[c] += 1
            else:
                occurrence[c] = 1
        for c in t:
            if c in occurrence:
                occurrence[c] -= 1
            else:
                return False
        for c in occurrence.values():
            if c != 0:
                return False
        return True

File: 389_Find_the_Difference/hash.cpp
class Solution {
  /*
   * Complexity: O(n)
   * Space: O(n)
   */
public:
    char findTheDifference(string s, string t) {
        unordered_map<char, int>occurrence;
        for(auto& c:s){
            occurrence[c] += 1;
        }
        for(auto& c:t){
            auto it = occurrence.find(c);
            if(it == occurrence.end()){
                return c;
            }
            else{
                if(--it -> second == 0){
                    occurrence.erase(it);
                }
            }
        }
    }
};
